sys/include/net/nanocoap_sock.h

```diff
@@ -279,12 +293,157 @@ int nanocoap_server_prepare_separate(nanocoap_server_response_ctx_t *ctx,
+ *
+ * @warning This function is only available when using the module
+ *          `nanocoap_server_separate`
+ *
+ * @param[in]   ctx     Context information for the CoAP response
+ * @param[in]   reply   I/O list containing the reply to send
+ *
+ * @retval      0   Success
+ * @retval      <0  negative errno code indicating the error
+ */
+int nanocoap_server_sendv_separate(const nanocoap_server_response_ctx_t *ctx,
+                                   const iolist_t *reply);
+
+/**
+ * @brief           Register an observer
+ * @param[in]       req_ctx         Request context belonging to @p req_pkt
+ * @param[in,out]   req_pkt         Request that contained the observe registration request
+ *
+ * @warning This depends on module `nanocoap_server_observe`
+ *
+ * @note    If the same endpoint already was registered on the same resource,
+ *          it will just update the token and keep the existing entry. This
+ *          way duplicate detection is not needed and we eagerly can reclaim
+ *          resources when a client lost state.
+ *
+ * @warning Preventing the same endpoint to registers more than once (using
+ *          different tokens) to the same resource deviates from RFC 7641.
+ *
+ * The deviation here is intentional. A server can receive a second registration
+ * from the same endpoint for the same resource for one of the following
+ * reasons:
+ *
+ * 1. Reaffirming the registration by using the same token again.
+ * 2. Loosing state on the client side.
+ * 3. A malicious client trying to exhaust resources.
+ * 4. The same resource has different representations depending on the
+ *    request. (E.g. `/.well-known/core` can yield a wildly different response
+ *    depending on filters provided via URI-Query Options.)
+ *
+ * For case 1 updating the registration is matching what the spec mandates.
+ * For two the old registration will not be of value for the client, and
+ * overwriting it makes more efficient use of network bandwidth and RAM.
+ * For 3 the deviation forces the adversary to send observe requests from
+ * different ports to exhaust resources, which is a very minor improvement.
+ * For 4 the deviation is a problem. However, the observe API does not allow to
+ * send out different notification messages for the same resource anyway, so
+ * case 4 cannot occur here.
+ *
+ * @retval  0           Success
+ * @retval  -ENOMEM     Not enough resources to register another observer
+ * @retval  <0          Negative errno code indicating error
+ */
+int nanocoap_register_observer(const coap_request_ctx_t *req_ctx, coap_pkt_t *req_pkt);
+
+/**
+ * @brief   Unregister an observer
+ * @param   req_ctx         Request context belonging to @p req_pkt
+ *
+ * @warning This depends on module `nanocoap_server_observe`
+ *
+ * @note    It is safe to call this multiple times, e.g. duplicate detection
+ *          is not needed for this.
+ */
+void nanocoap_unregister_observer(const coap_request_ctx_t *req_ctx);
```

Comment:
I think this is cutting to much corners.

A duplicated unregister request arriving late could also incorrectly remove a fresh entry in the list. I'll add a check for the token here as well.